codebasin/__main__.py

```diff
@@ -54,6 +55,68 @@ def _help_string(*lines: str, is_long=False, is_last=False):
     return result
 
 
+class MetaWarning:
+    """
+    A MetaWarning is used to represent multiple warnings, and provide suggested
+    actions to the user.
+    """
+
+    def __init__(self, regex: str, msg: str):
+        self.regex = re.compile(regex)
+        self.msg = msg
+        self._count = 0
+
+    def inspect(self, record):
+        if self.regex.search(record.msg):
+            self._count += 1
+
+    def warn(self):
+        if self._count == 0:
+            return
+        log.warning(self.msg.format(self._count))
+
+
+class WarningAggregator(logging.Filter):
+    """
+    Inspect warnings to generate meta-warnings and statistics.
+    """
+
+    def __init__(self):
+        self.meta_warnings = [
+            MetaWarning(".", "{} warnings generated during preprocessing."),
+            MetaWarning(
+                "user include",
+                "{} user include files could not be found.\n"
+                + "  These could contain important macros and includes.\n"
+                + "  Suggested solutions:\n"
+                + "  - Check that the file(s) exist in the code base.\n"
+                + "  - Check the include paths in the compilation database.\n"
+                + "  - Check if the include(s) should have used '<>'.",
+            ),
+            MetaWarning(
+                "system include",
+                "{} system include files could not be found.\n"
+                + "  These could define important feature macros.\n"
+                + "  Suggested solutions:\n"
+                + "  - Check that the file(s) exist on your system.\n"
+                + "  - Use .cbi/config to define system include paths.\n"
+                + "  - Use .cbi/config to define important macros.",
+            ),
+        ]
+
+    def filter(self, record):
```

Comment:
Missing type annotation on `record`

Reply:
Fixed in https://github.com/intel/code-base-investigator/pull/115/commits/9abcf0c545542fe2bdb853bea0c49267b4314e8a.